src/stream_inserter.cpp

```diff
@@ -13,34 +13,76 @@
 // limitations under the License.
 
 #include "stream_inserter.h"
-#include <memory>
-#include "grpcpp/client_context.h"
 
 namespace greptime {
-    
-using greptime::v1::RequestHeader;
-
-bool StreamInserter::Write(InsertRequests &insert_requests) {
-    RequestHeader request_header;
-    request_header.set_dbname(dbname);
-    // avoid repetitive construction and destruction
-    thread_local GreptimeRequest greptime_request;
-    greptime_request.mutable_header()->Swap(&request_header);
-    greptime_request.mutable_inserts()->Swap(&insert_requests);
-
-    while(true) {
-        if (writer->Write(greptime_request)) {
-            return true; // Write successful
-        } else {
-            if (channel->GetState(true) == GRPC_CHANNEL_TRANSIENT_FAILURE) {
-                context = std::make_shared<grpc::ClientContext>();
-                context->set_wait_for_ready(true);
-                writer = stub->HandleRequests(context.get(), &response);
-            } else {
-                return false; // Write failed
+
+#define BUFFER_SIZE 1000000
+
+void StreamInserter::Write(InsertRequest insert_request) {
+    std::unique_lock<std::mutex> lk(mtx);
+    cv.wait(lk, [this]{
+        return this->buffer.size() + 1 < (BUFFER_SIZE);
+    });
+    buffer.push(std::move(insert_request));
+    cv.notify_one();
+}
+
+void StreamInserter::WriteBatch(std::vector<InsertRequest> insert_request_vec) {
+    size_t cnt = insert_request_vec.size();
+    if (cnt > BUFFER_SIZE) {
+        std::cout << "fail: data too many" << std::endl;
+        return;
+    }
+    std::unique_lock<std::mutex> lk(mtx);
+    cv.wait(lk, [this, cnt]{
+        return this->buffer.size() + cnt < (BUFFER_SIZE);
+    }); 
+    for (auto &insert_request : insert_request_vec) {
+        buffer.push(std::move(insert_request));
+    }
+    cv.notify_one();
+}
+
+bool StreamInserter::Send(const GreptimeRequest &greptime_request) {
+    return writer->Write(greptime_request);
+}
+
+void StreamInserter::RunHandleRequest() {
+    while (true) {
+        std::unique_lock<std::mutex> lk(mtx);
+        cv.wait(lk, [this]{
+            return !this->buffer.empty() || !this->scheduler_thread_is_running;
+        });
+        if (buffer.empty() && !scheduler_thread_is_running) {
+            break;
+        }
+
+        if (!buffer.empty()) {
+            size_t batch_byte = 0;
+            #define BATCH_BYTE_LIMIT 2981328
+            InsertRequests insert_requests;
+            while (!buffer.empty() && batch_byte < BATCH_BYTE_LIMIT) {
+                if (batch_byte + buffer.front().ByteSizeLong() > BATCH_BYTE_LIMIT) {
+                    break;
+                } 
+                auto insert_request = std::move(buffer.front());
```

Comment:
Can we merge `buffer::front` with `buffer::pop`?
```suggestion
                auto insert_request = buffer.pop();
```

Reply:
std::queue::pop doesn't return value: [Why doesn't std::queue::pop return value.?](https://stackoverflow.com/questions/25035691/why-doesnt-stdqueuepop-return-value)